### PR TITLE
Assignment 8

### PR DESCRIPTION
**Pull request by Fervidusletum**
Finished

### CODE REVIEW

File: Assignment.Tests/SampleDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        [TestMethod]
+        public void GetUniqueSortedListOfStatesGivenCsvRowsTest_ReturnsSortedListOfStates_ListInAscendingOrder()
+        {
+            List<string> sutData = (List<string>) SUT.GetUniqueSortedListOfStatesGivenCsvRows();
+            IEnumerable<string> sorted = new List<string>(sutData) // copy data
+                .OrderBy(state => state).ToList();
+
+            Assert.IsTrue(Enumerable.SequenceEqual(sorted, sutData), message: "Not Sorted.");
```
Comment (Cybotive) [inline Assignment.Tests/SampleDataTests.cs]: I don't know if using OrderBy in your test is a good way to test the OrderBy in SampleData. What if both are wrong?

File: Assignment/Address.cs [hunk 2]
```diff
@@ -2,9 +2,9 @@
 {
     public class Address : IAddress
     {
-        public string StreetAddress { get; set; }
-        public string City { get; set; }
-        public string State { get; set; }
-        public string Zip { get; set; }
+        public string StreetAddress { get; set; } = "";
```
Comment (Cybotive) [inline Assignment/Address.cs]: I'm not sure what the point of blank string is. But a sidenote is that Mark pushed an updated version of the interfaces so they're not as redundant, I'm not sure if they're going to require you use the updated version though.

Comment (Fervidusletum) [reply Assignment/Address.cs]: I rebased onto that commit (896ede4c164489524dad26f45b16c2835359a754) before I really got started so I'm fine there, but I appreciate that you brought it up!

File: Assignment.Tests/SampleDataTests.cs (hunk 1 above)
Comment (Fervidusletum) [reply Assignment.Tests/SampleDataTests.cs]: Agreed, they were pretty poorly written tests.

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -11,32 +11,32 @@ public enum CsvColumn { Id, FirstName, LastName, Email, StreetAddress, City, Sta
...
-            // just some notes to future me here, don't mind them
-            List<string> states = (
-                from line in CsvRows
-                select line.Split(',')[(int)CsvColumn.State]) // transform the line into split strings
+            List<string> states =
+                // transform the line into tuple of states extracted from split strings
+                ( from line in CsvRows select line.Split(',')[(int) CsvColumn.State] )
                 .Distinct() // remove duplicates
                 .OrderBy((string state) => state) // compare by state string
                 .ToList(); // force query to execute
```
Comment (ashermancinelli) [inline Assignment/SampleData.cs]: Might be easier to understand if you consistently use one version of the query, eg
```c#
            List<string> states = CsvRows
                .Select(line => line.Split(',')[(int) CsvColumn.State])
                .Distinct()
                .OrderBy((string state) => state)
                .ToList();
```

Comment (ashermancinelli) [inline Assignment/SampleData.cs]: Might be easier to understand if you consistently use one version of the query, eg
```suggestion
            List<string> states = 
                // transform the line into tuple of states extracted from split strings
                CsvRows
                .Select(line => line.Split(',')[(int) CsvColumn.State])
                .Distinct() // remove duplicates
                .OrderBy((string state) => state) // compare by state string
                .ToList(); // force query to execute
```

File: Assignment/SampleData.cs [hunk 4]
```diff
@@ -11,32 +11,32 @@ public enum CsvColumn { Id, FirstName, LastName, Email, StreetAddress, City, Sta
         private const string _CSV_FILENAME = "People.csv";
```
Comment (ashermancinelli) [inline Assignment/SampleData.cs]: I think private variables are supposed to be in pascal case, eg 
```suggestion
         private const string _CsvFileName = "People.csv";
```

File: Assignment.Tests/SampleDataTests.cs [hunk 5]
```diff
@@ -50,7 +53,7 @@ public void GetUniqueSortedListOfStatesGivenCsvRows_ReturnsSortedListOfStates_Li
             bool hasDescending = sutData
                 .Where((string state, int i) => i < sutData.Count - 1) // ignore last element which would cause index OoB
                 .Select((string state, int i) => (state, sutData[i + 1])) // pair up elements sequentially
-                .Select(((string First, string Second) pair) => // compare pairs of elements to check ascending order
+                .Select(((string First, string Second) pair) => // compare elements of pair to check ascending order
                         string.CompareOrdinal(pair.First, pair.Second) < 0
                 ).Any(isAscending => isAscending == false); // find any result that indicates descending order
```
Comment (ashermancinelli) [inline Assignment.Tests/SampleDataTests.cs]:
```suggestion
                ).Any(isAscending => !isAscending); // find any result that indicates descending order
```

File: Assignment/Address.cs [hunk 6]
```diff
@@ -1,10 +1,78 @@
-﻿namespace Assignment
+﻿using System;
+
+namespace Assignment
 {
-    public class Address : IAddress
+    public class Address : IAddress, IEquatable<Address>, IComparable<Address>
```
Comment (MarkMichaelis) [inline Assignment/Address.cs]: Wow... above and beyond. :)

File: Assignment/SampleData.cs [hunk 7]
```diff
@@ -1,30 +1,112 @@
...
 namespace Assignment
 {
     public class SampleData : ISampleData
     {
+        public enum CsvColumn { Id, FirstName, LastName, EmailAddress, StreetAddress, City, State, Zip }
+        private const string _Csv_Filename = "People.csv";
+
         // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
+        private static IEnumerable<string> CsvDefault { get; } = File.ReadAllLines(_Csv_Filename).Skip(1).ToList();
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Why convert to ToList()?

**Pull request closed without merging**